File: java-spring/src/main/java/org/burgas/javaspring/service/IdentityService.java
package org.burgas.javaspring.service;

import jakarta.servlet.http.Part;
import org.burgas.javaspring.cache.KeyUtil;
import org.burgas.javaspring.cache.RedisHandler;
import org.burgas.javaspring.dto.identity.IdentityFullResponse;
import org.burgas.javaspring.dto.identity.IdentityRequest;
import org.burgas.javaspring.dto.identity.IdentityShortResponse;
import org.burgas.javaspring.dto.note.NoteFullResponse;
import org.burgas.javaspring.entity.identity.Identity;
import org.burgas.javaspring.mapper.IdentityMapper;
import org.burgas.javaspring.service.contract.CrudService;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Isolation;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@Transactional(readOnly = true, propagation = Propagation.NOT_SUPPORTED)
public class IdentityService implements
        CrudService<IdentityRequest, Identity, IdentityShortResponse, IdentityFullResponse>, RedisHandler<Identity> {

    private final IdentityMapper identityMapper;
    private final ImageService imageService;

    @Qualifier(value = "identityRedisTemplate")
    private final RedisTemplate<String, IdentityFullResponse> identityRedisTemplate;

    @Qualifier(value = "noteRedisTemplate")
    private final RedisTemplate<String, NoteFullResponse> noteRedisTemplate;

    public IdentityService(
            IdentityMapper identityMapper, ImageService imageService,
            RedisTemplate<String, IdentityFullResponse> identityRedisTemplate,
            RedisTemplate<String, NoteFullResponse> noteRedisTemplate
    ) {
        this.identityMapper = identityMapper;
        this.imageService = imageService;
        this.identityRedisTemplate = identityRedisTemplate;
        this.noteRedisTemplate = noteRedisTemplate;
    }

    @Override
    public Identity findEntity(UUID id) {
        return this.identityMapper.identityRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Identity not found"));
    }

    @Override
    public IdentityFullResponse findById(UUID id) {
        var identityKeyFormat = String.format(KeyUtil.IDENTITY_KEY, id);
        if (this.identityRedisTemplate.hasKey(identityKeyFormat)) {
            return this.identityRedisTemplate.opsForValue().get(identityKeyFormat);
        } else {
            var identity = this.findEntity(id);
            var identityFullResponse = this.identityMapper.toFullResponse(identity);
            this.identityRedisTemplate.opsForValue().set(identityKeyFormat, identityFullResponse);
            return identityFullResponse;
        }
    }

    @Override
    public List<IdentityShortResponse> findAll() {
        return this.identityMapper.identityRepository.findAll()
                .stream()
                .map(this.identityMapper::toShortResponse)
                .toList();
    }

    @Override
    @Transactional(
            isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
            rollbackFor = {Exception.class, Throwable.class, RuntimeException.class}
    )
    public void create(IdentityRequest request) {
        var identityFullResponse = this.identityMapper.toFullResponse(
                this.identityMapper.identityRepository.save(this.identityMapper.toEntity(request))
        );
        var identityKeyFormat = String.format(KeyUtil.IDENTITY_KEY, identityFullResponse.getId());
        this.identityRedisTemplate.opsForValue().set(identityKeyFormat, identityFullResponse);
    }

    @Override
    @Transactional(
            isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
            rollbackFor = {Exception.class, Throwable.class, RuntimeException.class}
    )
    public void update(IdentityRequest request) {
        if (request.getId() == null) throw new IllegalArgumentException("Identity Request id is null");
        Identity identity = this.identityMapper.identityRepository.save(this.identityMapper.toEntity(request));
        this.handleCache(identity);
    }

    @Override
    @Transactional(
            isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
            rollbackFor = {Exception.class, Throwable.class, RuntimeException.class}
    )
    public void delete(UUID id) {
        var identity = this.findEntity(id);
        this.identityMapper.identityRepository.delete(identity);
        this.handleCache(identity);
    }

    @Transactional(
            isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
            rollbackFor = {Exception.class, Throwable.class, RuntimeException.class}
    )
    public void changePassword(final IdentityRequest identityRequest) {
        if (identityRequest.getId() == null) throw new IllegalArgumentException("Identity Request id is null");
        if (identityRequest.getPassword() == null || identityRequest.getPassword().isEmpty())
            throw new IllegalArgumentException("Identity password is null or empty");

        var identity = this.findEntity(identityRequest.getId());
        if (this.identityMapper.passwordEncoder.matches(identityRequest.getPassword(), identity.getPassword()))
            throw new IllegalArgumentException("Passwords matched");

        identity.setPassword(this.identityMapper.passwordEncoder.encode(identityRequest.getPassword()));
        this.handleCache(identity);
    }

    @Transactional(
            isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
            rollbackFor = {Exception.class, Throwable.class, RuntimeException.class}
    )
    public void changeStatus(final IdentityRequest identityRequest) {
        if (identityRequest.getId() == null) throw new IllegalArgumentException("Identity Request id is null");
        if (identityRequest.getEnabled() == null) throw new IllegalArgumentException("Identity Request status is null");

        var identity = this.findEntity(identityRequest.getId());
        if (identityRequest.getEnabled() == identity.getEnabled())
            throw new IllegalArgumentException("Statuses matched");

        identity.setEnabled(identityRequest.getEnabled());
        this.handleCache(identity);
    }

    @Transactional(
            isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
            rollbackFor = {Exception.class, Throwable.class, RuntimeException.class}
    )
    public void uploadImage(final UUID identityId, final Part part) {
        var identity = this.findEntity(identityId);
        identity.setImage(this.imageService.uploadSingle(part));
        this.handleCache(identity);
    }

    @Transactional(
            isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
            rollbackFor = {Exception.class, Throwable.class, RuntimeException.class}
    )
    public void removeImages(final UUID identityId) {
        var identity = this.findEntity(identityId);
        var image = identity.getImage();
        if (image != null) {
            identity.setImage(null);
            this.imageService.remove(List.of(image.getId()));
            this.handleCache(identity);
        } else {
            throw new IllegalArgumentException("Identity image is null");
        }
    }

    @Override
    public void handleCache(Identity identity) {
        var identityKeyFormat = String.format(KeyUtil.IDENTITY_KEY, identity.getId());
        if (this.identityRedisTemplate.hasKey(identityKeyFormat)) this.identityRedisTemplate.delete(identityKeyFormat);
        if (!identity.getNotes().isEmpty()) {
            identity.getNotes().forEach(
                    note -> {
                        var noteKeyFormat = String.format(KeyUtil.NOTE_KEY, note.getId());
                        if (this.noteRedisTemplate.hasKey(noteKeyFormat)) this.noteRedisTemplate.delete(noteKeyFormat);
                    }
            );
        }
    }
}

File: kotlin-spring/src/main/kotlin/org/burgas/kotlinspring/service/IdentityService.kt
package org.burgas.kotlinspring.service

import jakarta.servlet.http.Part
import org.burgas.kotlinspring.entity.identity.Identity
import org.burgas.kotlinspring.entity.identity.IdentityFullResponse
import org.burgas.kotlinspring.entity.identity.IdentityRequest
import org.burgas.kotlinspring.entity.identity.IdentityShortResponse
import org.burgas.kotlinspring.entity.note.NoteFullResponse
import org.burgas.kotlinspring.kafka.CustomKafkaProducer
import org.burgas.kotlinspring.mapper.IdentityMapper
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Isolation
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional(readOnly = true, propagation = Propagation.NOT_SUPPORTED)
class IdentityService : CrudService<IdentityRequest, Identity, IdentityShortResponse, IdentityFullResponse> {

    private final val identityMapper: IdentityMapper
    private final val imageService: ImageService

    @Qualifier(value = "identityRedisTemplate")
    private final val identityRedisTemplate: RedisTemplate<String, IdentityFullResponse>

    @Qualifier(value = "noteRedisTemplate")
    private final val noteRedisTemplate: RedisTemplate<String, NoteFullResponse>

    private final val passwordEncoder: PasswordEncoder

    private final val customKafkaProducer: CustomKafkaProducer

    private val identityKey: String = "identityFullResponse::%s"
    private val noteKey: String = "noteFullResponse::%s"

    constructor(
        identityMapper: IdentityMapper,
        imageService: ImageService,
        identityRedisTemplate: RedisTemplate<String, IdentityFullResponse>,
        noteRedisTemplate: RedisTemplate<String, NoteFullResponse>,
        passwordEncoder: PasswordEncoder,
        customKafkaProducer: CustomKafkaProducer
    ) {
        this.identityMapper = identityMapper
        this.imageService = imageService
        this.identityRedisTemplate = identityRedisTemplate
        this.noteRedisTemplate = noteRedisTemplate
        this.passwordEncoder = passwordEncoder
        this.customKafkaProducer = customKafkaProducer
    }

    private fun handleCache(identity: Identity) {
        val identityKey = this.identityKey.format(identity.id)
        if (identityRedisTemplate.hasKey(identityKey)) identityRedisTemplate.delete(identityKey)

        if (!identity.notes.isEmpty()) {

            identity.notes.forEach { note ->
                val noteKey = this.noteKey.format(note.id)
                if (noteRedisTemplate.hasKey(noteKey)) noteRedisTemplate.delete(noteKey)
            }
        }
    }

    override fun findEntity(id: UUID): Identity {
        return this.identityMapper.identityRepository.findById(id)
            .orElseThrow { throw IllegalArgumentException("Identity not found") }
    }

    override fun findAll(): List<IdentityShortResponse> {
        return this.identityMapper.identityRepository.findAll()
            .map { this.identityMapper.toShortResponse(it) }
    }

    override fun findById(id: UUID): IdentityFullResponse {
        val identityKey = this.identityKey.format(id)
        val identityFullResponse = this.identityRedisTemplate.opsForValue().get(identityKey)

        if (identityFullResponse != null) {
            return identityFullResponse

        } else {
            val response = this.identityMapper.toFullResponse(this.findEntity(id))
            this.identityRedisTemplate.opsForValue().set(identityKey, response)
            return response
        }
    }

    @Transactional(
        isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
        rollbackFor = [Exception::class, Throwable::class, RuntimeException::class]
    )
    override fun create(request: IdentityRequest) {
        var identity = this.identityMapper.toEntity(request)
        identity = this.identityMapper.identityRepository.save(identity)
        val identityFullResponse = this.identityMapper.toFullResponse(identity)
        this.customKafkaProducer.sendIdentityFullResponse(identityFullResponse)
    }

    @Transactional(
        isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
        rollbackFor = [Exception::class, Throwable::class, RuntimeException::class]
    )
    override fun update(request: IdentityRequest) {
        if (request.id == null) throw IllegalArgumentException("Identity id is null")
        val identity = this.identityMapper.toEntity(request)
        this.identityMapper.identityRepository.save(identity)
        handleCache(identity)
    }

    @Transactional(
        isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
        rollbackFor = [Exception::class, Throwable::class, RuntimeException::class]
    )
    override fun delete(id: UUID) {
        val identity = this.findEntity(id)
        this.identityMapper.identityRepository.delete(identity)
        handleCache(identity)
    }

    @Transactional(
        isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
        rollbackFor = [Exception::class, Throwable::class, RuntimeException::class]
    )
    fun uploadImage(identityId: UUID, part: Part) {
        val identity = this.findEntity(identityId)
        val image = this.imageService.upload(part)
        identity.apply {
            this.image = image
        }
        this.identityMapper.identityRepository.save(identity)
        handleCache(identity)
    }

    @Transactional(
        isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
        rollbackFor = [Exception::class, Throwable::class, RuntimeException::class]
    )
    fun removeImage(identityId: UUID) {
        val identity = this.findEntity(identityId)
        val image = identity.image
        if (image != null) {
            identity.apply {
                this.image = null
            }
            this.identityMapper.identityRepository.save(identity)
            this.imageService.remove(image)
        } else {
            throw IllegalArgumentException("Identity image is null")
        }
        handleCache(identity)
    }

    @Transactional(
        isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
        rollbackFor = [Exception::class, Throwable::class, RuntimeException::class]
    )
    fun changePassword(identityRequest: IdentityRequest) {
        if (identityRequest.id == null) throw IllegalArgumentException("Identity id is null")
        if (identityRequest.password.isNullOrEmpty()) throw IllegalArgumentException("Identity password is null")

        val identity = this.findEntity(identityRequest.id)
        if (this.passwordEncoder.matches(identityRequest.password, identity.password)) {
            identity.apply {
                this.password = identityRequest.password
            }
            handleCache(identity)

        } else {
            throw IllegalArgumentException("Encoded and new password matches")
        }
    }

    @Transactional(
        isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED,
        rollbackFor = [Exception::class, Throwable::class, RuntimeException::class]
    )
    fun changeStatus(identityRequest: IdentityRequest) {
        if (identityRequest.id == null) throw IllegalArgumentException("Identity id is null")
        if (identityRequest.enabled == null) throw IllegalArgumentException("Identity status is null")

        val identity = this.findEntity(identityRequest.id)
        if (identityRequest.enabled == identity.enabled) {
            identity.apply {
                this.enabled = identityRequest.enabled
            }
            handleCache(identity)

        } else {
            throw IllegalArgumentException("Request and identity statuses matched")
        }
    }
}